### PR TITLE
Add winston-loki support

### LINKED ISSUE #7: Add winston-loki support
If configured by env, we should be enable to send logs to loki server.

### PR DESCRIPTION
**Pull request by nmlinaric**
Add `winston-loki` support + `winston-loki` typings.

Resolves #7

### CODE REVIEW

**Comment by mpetrunic** [comment]
@nmlinaric What's your issue?

File: @types/winston-loki/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import * as Transport from "winston-transport";
```
Comment (mpetrunic) [inline @types/winston-loki/index.d.ts:1]: import must be inside declare module

File: @types/winston-loki/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        json?: object;
+        batching?: boolean;
+        clearOnError?: boolean;
+        replaceOnError?: boolean;
+        replaceTimestamp?: boolean;
+        gracefulShutdown?: boolean;
+
+    }
+
+    export class LokiTransport extends Transport {
```
Comment (mpetrunic) [inline @types/winston-loki/index.d.ts:17]:
```suggestion
    export class LokiTransport extends Transport.TransportStream {
```

**Comment by nmlinaric** [comment]
> @nmlinaric What's your issue?

Not 100% sure if `winston-loki` typings are ok this way and having trouble with that `TransportStream`

File: .env.sample [hunk 3]
```diff
@@ -18,4 +18,5 @@ BACKEND_CONTAINER_NAME=fastify-starter-backend
 BACKEND_LOCAL_PORT=3000
 DB_CONTAINER_NAME=fastify-starter-db
 DB_LOCAL_PORT=3000
-GRAFANA_WHITELIST=127.0.0.1,127.0.0.2
+METRICS_IP_WHITELIST=127.0.0.1,127.0.0.2
+LOKI_LOGGING=false;
```
Comment (mpetrunic) [inline .env.sample:22]:
```suggestion
WINSTON_LOKI_URL=http://localhost:3000;
```
If not set don't add transport

File: src/services/logger/index.ts [hunk 4]
```diff
@@ -9,20 +11,28 @@ const format = winston.format.printf(({ level, message, label, timestamp, reques
...
+        winston.format.timestamp(),
+        winston.format.align(),
+        format
+    )
+  }),
+]
+
+if(process.env.LOKI_LOGGING ?? "") {
+  transportsConfig.push(new LokiTransport({
+    host: "http://localhost:3100",
```
Comment (mpetrunic) [inline src/services/logger/index.ts:26]: you should pass custom labels.

I think we are suing "label" as moduleName

**Comment by nmlinaric** [comment]
@mpetrunic I've added `GRAFANA_CONTAINER_NAME`, `GRAFANA_CONTAINER_PORT`, `WINSTON_LOKI_CONTAINER_NAME`, `WINSTON_LOKI_CONTAINER_PORT` `env` variables for setting them on the server in containers, are we going to need them or are they excess?

File: src/services/logger/index.ts [hunk 5]
```diff
@@ -1,28 +1,43 @@
...
+  })
+];
+
+if (process.env.WINSTON_LOKI_URL) {
+  transportsConfig.push(new LokiTransport({
+    host: process.env.WINSTON_LOKI_URL,
+    labels: {
+      app: process.env.WINSTON_LOKI_APP_NAME ?? "",
+      module: process.env.WINSTON_LOKI_MODULE,
+      level: process.env.WINSTON_LOKI_LEVEL
```
Comment (mpetrunic) [inline src/services/logger/index.ts:30]: nope, it has to be level set in message, not just hardcoded one

File: src/services/logger/index.ts [hunk 6]
```diff
@@ -1,28 +1,43 @@
...
+    )
+  })
+];
+
+if (process.env.WINSTON_LOKI_URL) {
+  transportsConfig.push(new LokiTransport({
+    host: process.env.WINSTON_LOKI_URL,
+    labels: {
+      app: process.env.WINSTON_LOKI_APP_NAME ?? "",
+      module: process.env.WINSTON_LOKI_MODULE,
```
Comment (mpetrunic) [inline src/services/logger/index.ts:29]: it has to be module attached to winston not env

File: src/services/logger/index.ts [hunk 7]
```diff
@@ -1,28 +1,43 @@
...
+    }
+  }));
+}
+
 export const logger = winston.createLogger({
   level: process.env.LOG_LEVEL || "debug",
 
   format: winston.format.json({}),
   defaultMeta: {
     label: "default"
```
Comment (mpetrunic) [inline src/services/logger/index.ts:40]: if you rename this to module and remove those static labels it might work out of box

**Review by mpetrunic — CHANGES_REQUESTED** [review]

**Comment by mpetrun5** [comment]
We are missing error logs here: https://github.com/NodeFactoryIo/node-ts-starter-fastify/blob/816798ce86cb1a19d7165bd1c8dbae6b3ef4a37c/src/services/logger/fastify.ts#L30.
Should be like this: https://github.com/NodeFactoryIo/aument-backend/blob/43fca025aeaf4dd2f8494231c5b2e451ad649422/src/services/logger/fastify.ts#L34

File: .env.sample [hunk 8]
```diff
@@ -18,4 +18,6 @@ BACKEND_CONTAINER_NAME=fastify-starter-backend
 BACKEND_LOCAL_PORT=3000
 DB_CONTAINER_NAME=fastify-starter-db
 DB_LOCAL_PORT=3000
-GRAFANA_WHITELIST=127.0.0.1,127.0.0.2
+METRICS_IP_WHITELIST=127.0.0.1,127.0.0.2
+WINSTON_LOKI_URL=http://loki:3100
```
Comment (mpetrunic) [inline .env.sample:22]: keep it empty in sample, so it  doesn't attempt to send it locally by default

File: .env.sample [hunk 9]
```diff
@@ -18,4 +18,6 @@ BACKEND_CONTAINER_NAME=fastify-starter-backend
 BACKEND_LOCAL_PORT=3000
 DB_CONTAINER_NAME=fastify-starter-db
 DB_LOCAL_PORT=3000
-GRAFANA_WHITELIST=127.0.0.1,127.0.0.2
+METRICS_IP_WHITELIST=127.0.0.1,127.0.0.2
```
Comment (mpetrunic) [inline .env.sample:21]: let it be * by default

File: src/services/logger/index.ts [hunk 10]
```diff
@@ -1,28 +1,44 @@
...
+      winston.format.align(),
+      format
+    )
+  })
+];
+
+if (process.env.WINSTON_LOKI_URL) {
+  transportsConfig.push(new LokiTransport({
+    host: process.env.WINSTON_LOKI_URL,
+    batching: false,
```
Comment (mpetrunic) [inline src/services/logger/index.ts:27]: env variable

**Review by mpetrunic — CHANGES_REQUESTED** [review]
Look good, few more comments and you can start integrating this in HP project (maybe you can merge it since they have same base commits)

**Review by mpetrunic — APPROVED** [review]

**Pull request merged**